flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java

```diff
@@ -151,23 +153,63 @@ int requestFloatingBuffers(int numRequired) throws IOException {
 				return numRequestedBuffers;
 			}
 
-			numRequiredBuffers = numRequired;
+			numRequiredBuffers += backlog;
+			numRequestedBuffers = internalRequestFloatingBuffers(numRequiredBuffers);
+			numRequiredBuffers -= numRequestedBuffers;
+		}
+		return numRequestedBuffers;
+	}
 
-			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
-				BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
-				Buffer buffer = bufferPool.requestBuffer();
-				if (buffer != null) {
-					bufferQueue.addFloatingBuffer(buffer);
-					numRequestedBuffers++;
-				} else if (bufferPool.addBufferListener(this)) {
-					isWaitingForFloatingBuffers = true;
-					break;
-				}
+	private int internalRequestFloatingBuffers(int numBuffersToRequest) throws IOException {
+		assert Thread.holdsLock(bufferQueue);
+
+		int numRequestedBuffers = 0;
+		while (numRequestedBuffers < numBuffersToRequest && !isWaitingForFloatingBuffers) {
+			BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
+			Buffer buffer = bufferPool.requestBuffer();
+			if (buffer != null) {
+				bufferQueue.addFloatingBuffer(buffer);
+				numRequestedBuffers++;
+			} else if (bufferPool.addBufferListener(this)) {
+				isWaitingForFloatingBuffers = true;
+				break;
 			}
 		}
 		return numRequestedBuffers;
 	}
 
+	public NettyMessage.ResumeConsumption resumeAndGetResumptionMessage(
```

Comment:
As mentioned in https://github.com/apache/flink/pull/11877/commits/708b2ceb97564084900c555d6f38e6ba1174d735#r424167698, this logic should be done inside `RemoteInputChannel`, because the `BufferManager` should not understand the specific logics unless buffer request/release.